app/models/host/discovered.rb

```diff
@@ -86,13 +86,17 @@ def populate_fields_from_facts(facts = self.facts_hash, type = 'puppet')
       # set location and organization
       if SETTINGS[:locations_enabled]
         self.location = Location.find_by_name(Setting[:discovery_location]) ||
+          Location.find_by_name(facts[Setting[:location_fact]]) ||
           subnet.try(:locations).try(:first) ||
-          Location.first
+          Location.find_by_name(Setting[:default_location]) ||
+          Location.first rescue Location.first
       end
       if SETTINGS[:organizations_enabled]
         self.organization = Organization.find_by_name(Setting[:discovery_organization]) ||
+          Organization.find_by_name(facts[Setting[:organization_fact]]) ||
           subnet.try(:organizations).try(:first) ||
-          Organization.first
+          Organization.find_by_name(Setting[:default_organization]) ||
+          Organization.first rescue Organization.first
       end
     else
       raise(::Foreman::Exception.new(N_("Unable to assign subnet, primary interface is missing IP address")))
```

Comment:
For this see http://theforeman.org/plugins/foreman_discovery/4.0/index.html#3.1.3OrganizationsandLocations